### ISSUE TITLE
install_github fails to correctly link dependencies

### ISSUE THREAD

**Issue by sheffien** [hadley/devtools#766]
Hi, 

I have a package hosted here: https://github.com/sheffien/IRangeKernels

If I install this package with `devtools:install_github("sheffien/IRangeKernels")`, I get this error:

```
Downloading github repo sheffien/IRangeKernels@master
Installing IRangeKernels
'/usr/lib/R/bin/R' --vanilla CMD INSTALL  \
  '/tmp/Rtmp6GpfdN/devtools194d1c2adafa/sheffien-IRangeKernels-8e77a03'  \
  --library='/home/nsheffield/R' --install-tests 

* installing *source* package ‘IRangeKernels’ ...
** libs
gcc -std=gnu99 -I/usr/share/R/include -DNDEBUG   -I"/home/nsheffield/R/IRanges/include"   -fpic  -g -O2 -fstack-protector --param=ssp-buffer-size=4 -Wformat -Werror=format-security -D_FORTIFY_SOURCE=2 -g  -c R_init_IRangeKernels.c -o R_init_IRangeKernels.o
In file included from IRangeKernels.h:1:0,
                 from R_init_IRangeKernels.c:1:
/home/nsheffield/R/IRanges/include/IRanges_defines.h:18:31: fatal error: S4Vectors_defines.h: No such file or directory
 #include "S4Vectors_defines.h"
```

If I instead clone the Repo and install locally, it loads fine:
```
install.packages("github/IRangeKernels", repos=NULL)
Installing package into ‘/home/nsheffield/R’
(as ‘lib’ is unspecified)
> tools:::.install_packages()
* installing *source* package ‘IRangeKernels’ ...
** libs
make: Nothing to be done for `all'.
installing to /home/nsheffield/R/IRangeKernels/libs
** R
** preparing package for lazy loading
** help
*** installing help indices
** building package indices
** testing if installed package can be loaded
> tools:::.test_load_package('IRangeKernels', '/home/nsheffield/R')
> 
* DONE (IRangeKernels)
```

Is this a bug in install_github() ? Or did I miss something?

```
devtools::session_info()
Session info -------------------------------------------------------------------
 setting  value                       
 version  R version 3.2.0 (2015-04-16)
 system   x86_64, linux-gnu           
 ui       X11                         
 language en_US                       
 collate  en_US.UTF-8                 
 tz       <NA>                        

Packages -----------------------------------------------------------------------
 package       * version  date       source        
 BiocInstaller   1.19.3   2015-04-21 Bioconductor  
 bitops        * 1.0-6    2013-08-17 CRAN (R 3.2.0)
 devtools        1.7.0    2015-01-17 CRAN (R 3.2.0)
 httr          * 0.6.1    2015-01-01 CRAN (R 3.2.0)
 RCurl         * 1.95-4.5 2014-12-28 CRAN (R 3.2.0)
 rstudioapi    * 0.3.1    2015-04-07 CRAN (R 3.2.0)
 stringr       * 0.6.2    2012-12-06 CRAN (R 3.2.0)
```

**Comment by hadley** [comment]
Can you figure out where `"S4Vectors_defines.h"` is located? It might be that `install_github()` isn't setting a path quite right.  Do you have a custom `.Rprofile`?

**Comment by sheffien** [comment]
`"S4Vectors_defines.h"` is in S4Vectors/include.  S4Vectors is included in the IRanges LinkingTo field.

The `IRanges/include/IRanges_defines.h` file refers to the `"S4Vectors/include/S4Vectors_defines.h"` file.

I do have a custom .Rprofile, but if I remove it, I get the same error, it's not related to any setting therein. Anyway, it doesn't do much: 

```
local({r <- getOption("repos");
       r["CRAN"] <- "http://cran.at.r-project.org";   #AUSTRIA mirror
       options(repos=r)})
```
and then sets some options and defines a few functions

**Comment by hadley** [comment]
Hmmmmmmmmm. What happens if you clone then use `install_local()`?

**Comment by sheffien** [comment]
That appears to work:

```
install_local("rpack/IRangeKernels/")                                                                 │
Installing IRangeKernels                                                                                │
'/usr/lib/R/bin/R' --vanilla CMD INSTALL  \                                                             │
  '/tmp/RtmppbNynh/file8cc53934bb/IRangeKernels'  \                                                     │
  --library='/home/nsheffield/R' --install-tests                                                        │
                                                                                                        │
* installing *source* package ‘IRangeKernels’ ...                                                       │
** libs                                                                                                 │
make: Nothing to be done for `all'.                                                                     │
installing to /home/nsheffield/R/IRangeKernels/libs                                                     │
** R                                                                                                    │
** preparing package for lazy loading                                                                   │
** help                                                                                                 │
*** installing help indices                                                                             │
** building package indices                                                                             │
** testing if installed package can be loaded                                                           │
* DONE (IRangeKernels)
```

**Comment by hadley** [comment]
Can you try again with the dev version? I can no longer reproduce the problem.

**Comment by sheffien** [comment]
The dev version (of devtools) seems to have broken install_github() for me, for any package:

```
DONE (devtools)
Reloading installed devtools
> devtools::install_github("hadley/devtools")
Downloading github repo hadley/devtools@master
Error: is.config(config) is not TRUE
> devtools::install_github("sheffien/IRangesKernels")
Downloading github repo sheffien/IRangesKernels@master
Error: is.config(config) is not TRUE
```

```
devtools::session_info()
Session info -------------------------------------------------------------------
 setting  value                       
 version  R version 3.2.0 (2015-04-16)
 system   x86_64, linux-gnu           
 ui       X11                         
 language en_US                       
 collate  en_US.UTF-8                 
 tz       <NA>                        

Packages -----------------------------------------------------------------------
 package       * version    date       source                          
 BiocInstaller   1.19.4     2015-04-24 Bioconductor                    
 bitops        * 1.0-6      2013-08-17 CRAN (R 3.2.0)                  
 devtools        1.7.0.9000 2015-04-29 Github (hadley/devtools@dee7b74)
 digest        * 0.6.8      2014-12-31 CRAN (R 3.2.0)                  
 httr          * 0.6.1.9000 2015-04-29 Github (hadley/httr@75aca2b)    
 knitr         * 1.10       2015-04-23 CRAN (R 3.2.0)                  
 memoise       * 0.2.1      2014-04-22 CRAN (R 3.2.0)                  
 RCurl         * 1.95-4.6   2015-04-24 CRAN (R 3.2.0)                  
 rstudioapi    * 0.3.1      2015-04-07 CRAN (R 3.2.0)                  
 stringr       * 0.6.2      2012-12-06 CRAN (R 3.2.0)                
```

**Comment by hadley** [comment]
Sorry, try again now

**Comment by sheffien** [comment]
Ok, this solved the config error, but I'm still getting the original includes error:

```
> devtools::install_local("rpack/IRangeKernels")
Installing IRangeKernels
'/usr/lib/R/bin/R' --no-site-file --no-environ --no-save --no-restore CMD  \
  INSTALL '/tmp/RtmpmAMxQb/file387a7522d7d7/IRangeKernels'  \
  --library='/home/nsheffield/R' --install-tests 

* installing *source* package ‘IRangeKernels’ ...
** libs
make: Nothing to be done for `all'.
installing to /home/nsheffield/R/IRangeKernels/libs
** R
** preparing package for lazy loading
** help
*** installing help indices
** building package indices
** testing if installed package can be loaded
* DONE (IRangeKernels)
> devtools::install_github("sheffien/IRangeKernels")
Downloading github repo sheffien/IRangeKernels@master
Installing IRangeKernels
'/usr/lib/R/bin/R' --no-site-file --no-environ --no-save --no-restore CMD  \
  INSTALL  \
  '/tmp/RtmpmAMxQb/devtools387a678b6811/sheffien-IRangeKernels-8e77a03'  \
  --library='/home/nsheffield/R' --install-tests 

* installing *source* package ‘IRangeKernels’ ...
** libs
gcc -std=gnu99 -I/usr/share/R/include -DNDEBUG   -I"/home/nsheffield/R/IRanges/include"   -fpic  -g -O2 -fstack-protector --param=ssp-buffer-size=4 -Wformat -Werror=format-security -D_FORTIFY_SOURCE=2 -g  -c R_init_IRangeKernels.c -o R_init_IRangeKernels.o
In file included from IRangeKernels.h:1:0,
                 from R_init_IRangeKernels.c:1:
/home/nsheffield/R/IRanges/include/IRanges_defines.h:18:31: fatal error: S4Vectors_defines.h: No such file or directory
 #include "S4Vectors_defines.h"
                               ^
compilation terminated.
make: *** [R_init_IRangeKernels.o] Error 1
ERROR: compilation failed for package ‘IRangeKernels’
* removing ‘/home/nsheffield/R/IRangeKernels’
* restoring previous ‘/home/nsheffield/R/IRangeKernels’
Error: Command failed (1)
> devtools::session_info()
Session info -------------------------------------------------------------------
 setting  value                       
 version  R version 3.2.0 (2015-04-16)
 system   x86_64, linux-gnu           
 ui       X11                         
 language en_US                       
 collate  en_US.UTF-8                 
 tz       <NA>                        

Packages -----------------------------------------------------------------------
 package       * version    date       source                          
 BiocInstaller   1.19.4     2015-04-24 Bioconductor                    
 bitops        * 1.0-6      2013-08-17 CRAN (R 3.2.0)                  
 devtools      * 1.7.0.9000 2015-04-29 Github (hadley/devtools@1a80e3c)
 httr          * 0.6.1.9000 2015-04-29 Github (hadley/httr@75aca2b)    
 RCurl         * 1.95-4.6   2015-04-24 CRAN (R 3.2.0)                  
 rstudioapi    * 0.3.1      2015-04-07 CRAN (R 3.2.0)                  
 stringr       * 0.6.2      2012-12-06 CRAN (R 3.2.0)           
```

**Comment by hadley** [comment]
I wonder if it's because you have an old version of IRanges. Can you try updating with `BiocInstaller::biocLite()`?

**Comment by sheffien** [comment]
I'm using the devel version, and IRanges was just a couple of days old:

```
Packages -----------------------------------------------------------------------
 package       * version    date       source                          
 BiocGenerics    0.15.0     2015-04-21 Bioconductor                    
 BiocInstaller   1.19.4     2015-04-24 Bioconductor                    
 bitops        * 1.0-6      2013-08-17 CRAN (R 3.2.0)                  
 devtools      * 1.7.0.9000 2015-04-29 Github (hadley/devtools@1a80e3c)
 httr          * 0.6.1.9000 2015-04-29 Github (hadley/httr@75aca2b)    
 IRanges         2.3.2      2015-04-24 Bioconductor                    
 RCurl         * 1.95-4.6   2015-04-24 CRAN (R 3.2.0)                  
 rstudioapi    * 0.3.1      2015-04-07 CRAN (R 3.2.0)                  
 S4Vectors       0.7.0      2015-04-21 Bioconductor                    
 stringr       * 0.6.2      2012-12-06 CRAN (R 3.2.0) 
```
After `BiocInstaller::biocLite()`:
```
Packages -----------------------------------------------------------------------
 package       * version    date       source                          
 BiocGenerics    0.15.0     2015-04-21 Bioconductor                    
 BiocInstaller   1.19.4     2015-04-24 Bioconductor                    
 bitops        * 1.0-6      2013-08-17 CRAN (R 3.2.0)                  
 devtools      * 1.7.0.9000 2015-04-29 Github (hadley/devtools@1a80e3c)
 httr          * 0.6.1.9000 2015-04-29 Github (hadley/httr@75aca2b)    
 IRanges         2.3.3      2015-04-29 Bioconductor                    
 RCurl         * 1.95-4.6   2015-04-24 CRAN (R 3.2.0)                  
 rstudioapi    * 0.3.1      2015-04-07 CRAN (R 3.2.0)                  
 S4Vectors       0.7.0      2015-04-21 Bioconductor                    
 stringr       * 0.6.2      2012-12-06 CRAN (R 3.2.0)       
```

But still no dice:
```
devtools::install_github("sheffien/IRangeKernels")
Downloading github repo sheffien/IRangeKernels@master
Installing IRangeKernels
'/usr/lib/R/bin/R' --no-site-file --no-environ --no-save --no-restore CMD  \
  INSTALL  \
  '/tmp/RtmpmAMxQb/devtools387a151ac68b/sheffien-IRangeKernels-8e77a03'  \
  --library='/home/nsheffield/R' --install-tests 

* installing *source* package ‘IRangeKernels’ ...
** libs
gcc -std=gnu99 -I/usr/share/R/include -DNDEBUG   -I"/home/nsheffield/R/IRanges/include"   -fpic  -g -O2 -fstack-protector --param=ssp-buffer-size=4 -Wformat -Werror=format-security -D_FORTIFY_SOURCE=2 -g  -c R_init_IRangeKernels.c -o R_init_IRangeKernels.o
In file included from IRangeKernels.h:1:0,
                 from R_init_IRangeKernels.c:1:
/home/nsheffield/R/IRanges/include/IRanges_defines.h:18:31: fatal error: S4Vectors_defines.h: No such file or directory
 #include "S4Vectors_defines.h"
                               ^
compilation terminated.
make: *** [R_init_IRangeKernels.o] Error 1
ERROR: compilation failed for package ‘IRangeKernels
```


Anyway I have the same problem on the previous version of R with the release version of bioconductor, on a different machine:

```
> devtools::install_github("sheffien/IRangeKernels")
Downloading github repo sheffien/IRangeKernels@master
Installing IRangeKernels
'/cm/shared/apps/R/3.1.2/lib64/R/bin/R' --vanilla CMD INSTALL  \
  '/tmp/RtmpFhPMyp/devtools2c1359077e2b/sheffien-IRangeKernels-8e77a03'  \
  --library='/home/nsheffield/.R' --install-tests 

* installing *source* package ‘IRangeKernels’ ...
** libs
gcc -std=gnu99 -I/cm/shared/apps/R/3.1.2/lib64/R/include -DNDEBUG  -I/usr/local/include -I"/home/nsheffield/.R/IRanges/include"   -fpic  -g -O2  -c R_init_IRangeKernels.c -o R_init_IRangeKernels.o
In file included from IRangeKernels.h:1:0,
                 from R_init_IRangeKernels.c:1:
/home/nsheffield/.R/IRanges/include/IRanges_defines.h:18:31: fatal error: S4Vectors_defines.h: No such file or directory
 #include "S4Vectors_defines.h"
                               ^
compilation terminated.
make: *** [R_init_IRangeKernels.o] Error 1
ERROR: compilation failed for package ‘IRangeKernels’
* removing ‘/home/nsheffield/.R/IRangeKernels’
* restoring previous ‘/home/nsheffield/.R/IRangeKernels’
Session info -------------------------------------------------------------------
 setting  value                       
 version  R version 3.1.2 (2014-10-31)
 system   x86_64, linux-gnu           
 ui       X11                         
 language (EN)                        
 collate  en_US.UTF-8                 
 tz       <NA>                        

Packages -----------------------------------------------------------------------
 package       * version    date       source                          
 BiocGenerics    0.12.1     2015-03-04 Bioconductor                    
 BiocInstaller   1.16.4     2015-04-24 Bioconductor                    
 bitops        * 1.0-6      2013-08-17 CRAN (R 3.1.2)                  
 devtools      * 1.7.0.9000 2015-04-29 Github (hadley/devtools@14cca3f)
 httr          * 0.6.1      2015-01-01 CRAN (R 3.1.2)                  
 IRanges         2.0.1      2015-03-09 Bioconductor                    
 RCurl         * 1.95-4.5   2014-12-28 CRAN (R 3.1.2)                  
 rstudioapi    * 0.3.1      2015-04-07 CRAN (R 3.1.2)                  
 S4Vectors       0.4.0      2015-03-09 Bioconductor                    
 stringr       * 0.6.2      2012-12-06 CRAN (R 3.1.0)
```

**Comment by hadley** [comment]
i can't reproduce locally, and I have very similar versions. I can't imagine what the problem would be, since `install_github()` is basically just downloads the package and then installs it (with a thin wrapper around `install.packages()`.

**Comment by hadley** [comment]
Comparing the compiler lines from you:

```
gcc -std=gnu99 -I/usr/share/R/include -DNDEBUG   -I"/home/nsheffield/R/IRanges/include"   -fpic  -g -O2 -fstack-protector --param=ssp-buffer-size=4 -Wformat -Werror=format-security -D_FORTIFY_SOURCE=2 -g  -c R_init_IRangeKernels.c -o R_init_IRangeKernels.o
```

and me

```
ccache clang -Qunused-arguments  -I/Library/Frameworks/R.framework/Resources/include -DNDEBUG  -I/usr/local/include -I/usr/local/include/freetype2 -I/opt/X11/include -I"/Users/hadley/R/IRanges/include"   -fPIC  -Wall -mtune=core2 -g -O2  -c R_init_IRangeKernels.c -o R_init_IRangeKernels.o
```

I don't see anything obviously wrong

**Comment by sheffien** [comment]
I can reproduce it across 4 different instances here now (all Linux), virtual machines, cluster, local computers. If you don't have the same problem it must have something to do with environment or OS or something. I wonder if it's a linux-only thing.

Anyway, I'll let you know if I figure anything else out.

**Comment by hadley** [comment]
I'm going to close for now since I can't reproduce. If anything else comes to light, please let me know.

**Comment by sheffien** [comment]
I traced this problem to something outside `install_github`... Basically, I hadn't accounted for the fact that when I cloned from github, I already had a `.so` file in there, since I was cloning into an existing repo. 

So `install_github` was showing the problem because it was actually pulling a fresh clone, which thus lacked the `.so` file. 

I still haven't solved the problem, but it's not with `install_github()` as I had thought. My mistake, and sorry to lead you down a wild goose chase.
